File: csharp/glp_link/primitives/LinkRequestKernel.cs
using GlpRuntime.Link.Reliability;
using GlpRuntime.Link.Seam;
using GlpRuntime.Multiagent;
using GlpRuntime.Runtime;

namespace GlpRuntime.Link.Primitives;

/// <summary>
/// The <c>'_link_request'/5</c> body kernel (feature 025, T033) — the host body of
/// <c>request_link(LinkId, ToPeer, ch(In?, Out), Faults?) :- ground(LinkId?), ground(ToPeer?) | '_link_request'(LinkId?, ToPeer?, In, Out?, Faults).</c>
/// (path-B establishment, FR-002). It is the CONNECTOR half of the handshake: it
/// connects to the peer's rendezvous over the transport (OQ-A3), emits the in-band
/// <c>request(LinkId, FromPeer)</c> token as a raw out-of-band frame, then establishes
/// the data link through the shared <see cref="LinkEstablish.WireEstablishedLink"/>
/// core — so the resulting link is indistinguishable from a listen/connect one (R-5).
/// </summary>
/// <remarks>
/// Arguments (ground-guarded by the wrapper): <c>args[0]</c> LinkId, <c>args[1]</c>
/// ToPeer (the peer AgentId; validated ground), <c>args[2]</c> In writer,
/// <c>args[3]</c> Out? reader, <c>args[4]</c> Faults writer. The token's
/// <c>FromPeer</c> is a base placeholder (<c>requester</c>) — authenticated origin
/// binding is FR-026/T075.
/// </remarks>
public static class LinkRequestKernel
{
    private const string Who = "'_link_request'/5";

    public static BodyKernelResult LinkRequest(GlpRuntimeEngine rt, IReadOnlyList<object?> args, LinkRuntime link)
    {
        if (args.Count != 5)
            return LinkEstablish.Abort(Who, $"expected 5 arguments, got {args.Count}");

        var heap = rt.Heap;
        LinkId id;
        try
        {
            id = LinkTerms.ParseLinkId(LinkTerms.GroundResolve(heap, (Term)args[0]!));
        }
        catch (ArgumentException ex)
        {
            return LinkEstablish.Abort(Who, $"arg 1 (LinkId): {ex.Message}");
        }
        if (heap.Dereference((Term)args[1]!) is VarRef)
            return LinkEstablish.Abort(Who, "arg 2 (ToPeer) must be a ground peer id");

        // Verify-before-act (FR-008, P1 fix 2026-07-13): gate BEFORE opening the QUIC connection or
        // shipping the request token. The gate keys only on the (already-parsed) ground LinkId, so a
        // refused requester never establishes a transport connection or exchanges the pre-data
        // handshake. WireEstablishedLink below is told preGated:true so the outcome is recorded once.
        if (LinkEstablish.CapabilityRefusal(link, id, Who) is BodyKernelResult refusal)
            return refusal;

        // Connect to the peer's rendezvous (blocks until the request_listener is there;
        // bounded by ConnectTimeout), then ship the in-band request token. The endpoint
        // is captured so the establish-core adopts the SAME connection it handshook on.
        ILinkEndpoint endpoint;
        try
        {
            var transport = link.Transports.Select(id.Scheme);
            var opts = LinkOptions.Default;
            using var cts = new CancellationTokenSource(opts.ConnectTimeout);
            endpoint = transport.ConnectAsync(id.Scheme, id.Endpoint, opts, cts.Token).GetAwaiter().GetResult();

            var token = LinkTerms.RequestToken(id, new ConstTerm("requester"));
            byte[] payload = new PayloadSerializer(string.Empty).SerializeAgentMessage(token);
            // Raw out-of-band frame (messageId 0, NOT via the link Sequencer) so the data
            // sequence still starts at 0; the listener consumes it before the data pump.
            foreach (var frame in FrameCodec.Encode(payload, messageId: 0u))
                endpoint.SendBytesAsync(frame).GetAwaiter().GetResult();
        }
        // ANY connect/handshake failure fails CLOSED GRACEFULLY (Abort), never an uncaught crash —
        // e.g. a "quic" link on a host without MsQuic throws PlatformNotSupportedException, and a
        // connect-timeout throws OperationCanceledException; neither was in the old narrow filter
        // (codexreview 20260713T110357Z P2). Same root fix as the LinkEstablish establishment catch.
        catch (Exception ex)
        {
            return LinkEstablish.Abort(Who, $"connect/handshake failed for {id}: {ex.Message}");
        }

        var result = LinkEstablish.WireEstablishedLink(
            rt, link, id, () => endpoint, args[2], args[3], args[4], Who, preGated: true);
        // 061 US4: the requester is the CONNECTOR half of the handshake — recorded for
        // snapshot section 0x09 re-establishment (contracts/snapshot-store.md).
        if (result == BodyKernelResult.Success && link.Links.TryGet(id, out var handle))
            handle.Role = LinkRole.Connector;
        return result;
    }
}

File: glp_runtime/lib/link/primitives/link_request_kernel.dart
import 'dart:typed_data';

import '../../multiagent/payload_serializer.dart';
import '../../runtime/body_kernels.dart' show BodyKernelResult;
import '../../runtime/runtime.dart';
import '../../runtime/terms.dart';
import '../reliability/frame_codec.dart';
import '../seam/i_link_endpoint.dart';
import '../seam/link_id.dart';
import '../seam/link_options.dart';
import 'link_establish.dart';
import 'link_runtime.dart';
import 'link_terms.dart';

/// The `'_link_request'/5` body kernel (feature 025, T033) — the host body of
/// `request_link(LinkId, ToPeer, ch(In?, Out), Faults?) :- ground(LinkId?), ground(ToPeer?) | '_link_request'(LinkId?, ToPeer?, In, Out?, Faults).`
/// (path-B establishment, FR-002). It is the CONNECTOR half of the handshake: it
/// connects to the peer's rendezvous over the transport (OQ-A3), emits the in-band
/// `request(LinkId, FromPeer)` token as a raw out-of-band frame, then establishes the
/// data link through the shared [LinkEstablish.wireEstablishedLink] core — so the
/// resulting link is indistinguishable from a listen/connect one (R-5).
///
/// Arguments (ground-guarded by the wrapper): `args[0]` LinkId, `args[1]` ToPeer (the
/// peer AgentId; validated ground), `args[2]` In writer, `args[3]` Out? reader,
/// `args[4]` Faults writer. The token's `FromPeer` is a base placeholder
/// (`requester`) — authenticated origin binding is FR-026/T075.
///
/// ASYNC-AWARE (Dart single-isolate adaptation of the C# blocking connect +
/// `GetAwaiter().GetResult()` handshake): the connect-and-ship-token is folded INTO
/// the `Future<ILinkEndpoint> Function()` establish closure handed to
/// [LinkEstablish.wireEstablishedLink], which wires the heap cursors + registry
/// SYNCHRONOUSLY and kicks the connect off without awaiting. A connect/handshake
/// FAILURE surfaces as a transport fault on the establishment `Faults` monitor (the
/// establish-core's onError path; FR-008) rather than as a synchronous Abort, since
/// the isolate cannot block on the rendezvous to observe it before returning.
class LinkRequestKernel {
  LinkRequestKernel._();

  static const String _who = "'_link_request'/5";

  static BodyKernelResult linkRequest(
      GlpRuntime rt, List<Object?> args, LinkRuntime link) {
    if (args.length != 5) {
      return LinkEstablish.abort(_who, 'expected 5 arguments, got ${args.length}');
    }

    final heap = rt.heap;
    LinkId id;
    try {
      id = LinkTerms.parseLinkId(LinkTerms.groundResolve(heap, args[0]! as Term));
    } on ArgumentError catch (ex) {
      return LinkEstablish.abort(_who, 'arg 1 (LinkId): ${ex.message}');
    }
    if (heap.dereference(args[1]! as Term) is VarRef) {
      return LinkEstablish.abort(_who, 'arg 2 (ToPeer) must be a ground peer id');
    }

    return LinkEstablish.wireEstablishedLink(
        rt, link, id, () => _connectAndHandshake(link, id),
        args[2], args[3], args[4], _who);
  }

  /// Connect to the peer's rendezvous (bounded by [LinkOptions.connectTimeout]),
  /// ship the in-band `request(LinkId, FromPeer)` token as a raw out-of-band frame
  /// (messageId 0, NOT via the link sequencer, so the data sequence still starts at
  /// 0; the listener consumes it before the data pump), and adopt the SAME connection
  /// the handshake rode for the data link.
  static Future<ILinkEndpoint> _connectAndHandshake(
      LinkRuntime link, LinkId id) async {
    final transport = link.transports.select(id.scheme);
    final opts = LinkOptions.default_;
    final cancel = Future<void>.delayed(opts.connectTimeout);
    final endpoint =
        await transport.connect(id.scheme, id.endpoint, opts, cancel: cancel);

    final token = LinkTerms.requestToken(id, ConstTerm('requester'));
    final List<int> payload = PayloadSerializer('').serializeAgentMessage(token);
    for (final Uint8List frame in FrameCodec.encode(payload, 0)) {
      await endpoint.sendBytes(frame);
    }
    return endpoint;
  }
}
